src/Microsoft.Health.TaskManagement/JobHosting.cs

```diff
@@ -37,46 +37,50 @@ public JobHosting(IQueueClient queueClient, IJobFactory jobFactory, ILogger<JobH
 
         public int JobHeartbeatTimeoutThresholdInSeconds { get; set; } = Constants.DefaultJobHeartbeatTimeoutThresholdInSeconds;
 
-        public int JobHeartbeatIntervalInSeconds { get; set; } = Constants.DefaultJobHeartbeatIntervalInSeconds;
+        public double JobHeartbeatIntervalInSeconds { get; set; } = Constants.DefaultJobHeartbeatIntervalInSeconds;
 
         public async Task ExecuteAsync(byte queueType, string workerName, CancellationTokenSource cancellationTokenSource, bool useHeavyHeartbeats = false)
         {
-            var runningJobs = new List<Task>();
+            var workers = new List<Task>();
 
-            while (!cancellationTokenSource.Token.IsCancellationRequested)
+            // parallel dequeue
+            for (var thread = 0; thread < MaxRunningJobCount; thread++)
             {
-                if (runningJobs.Count >= MaxRunningJobCount)
+                workers.Add(Task.Run(async () =>
                 {
-                    _ = await Task.WhenAny(runningJobs.ToArray());
-                    runningJobs.RemoveAll(t => t.IsCompleted);
-                }
+                    // random delay to avoid convoys
+                    await Task.Delay(TimeSpan.FromSeconds(RandomNumberGenerator.GetInt32(100) / 100.0 * PollingFrequencyInSeconds));
 
-                JobInfo nextJob = null;
-                if (_queueClient.IsInitialized())
-                {
-                    try
+                    while (!cancellationTokenSource.Token.IsCancellationRequested)
                     {
-                        nextJob = await _queueClient.DequeueAsync(queueType, workerName, JobHeartbeatTimeoutThresholdInSeconds, cancellationTokenSource.Token);
+                        JobInfo nextJob = null;
+                        if (_queueClient.IsInitialized())
+                        {
+                            try
+                            {
+                                nextJob = await _queueClient.DequeueAsync(queueType, workerName, JobHeartbeatTimeoutThresholdInSeconds, cancellationTokenSource.Token);
+                            }
+                            catch (Exception ex)
+                            {
+                                _logger.LogError(ex, "Failed to dequeue new job.");
+                            }
+                        }
+
+                        if (nextJob != null)
+                        {
+                            await ExecuteJobAsync(nextJob, useHeavyHeartbeats);
+                        }
+                        else
+                        {
+                            await Task.Delay(TimeSpan.FromSeconds(PollingFrequencyInSeconds), cancellationTokenSource.Token);
+                        }
```

Comment:
For worker threads, we need to make sure no exception for all operations, I believe redundant try catch is necessary.

Reply:
Are you talking about try/catch around ExecuteJobAsync in ExecuteAsync method?
If so, then ExecuteJobAsync is already protected by try/catch logic. I would understand the need for redundant try/catch in ExecuteAsync if ExecuteJobAsync was a method from a different class. But ExecuteJobAsync method is private, so its sole purpose is to be right for JobHosting class. 
Adding redundant try/catch in this particular case does not make sense to me.

Reply:
I mean there could be some other operations like _queueClient.IsInitialized() and possible further more.

Reply:
@tongwu-msft Did these problems exist in the original code (before PR)?

Reply:
Previous if there's any exception throw, the JobHosting would stop and the process for PaaS would be able to restart to recover, if we wait for all workers complete then my concern if any exception in 1 worker would only stop this thread and would not recover.